MudTreeViewItem: Add 'BodyContent' which renders instead of text, end text and end icon.

rpc-scandinavia opened the pull request:
## Description
The 'Content' render fragment replaces the entire tree view item, which mean that you have to handle the toggle button and logic your self.

I needed to use the nice logic of the tree view (loading, expanding/collapsing etc.), but provide a custom body.

I also needed a way to reload the tree view item.

## How Has This Been Tested?
Uses it in mu project.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

mikes-gh commented:
This doesn't build and would need tests for the new API added.  But before you spend more time on this we would need to see what the new API brings. @tungi52

rpc-scandinavia commented:
Interesting with a new API in sight.

I use this in my project, with "MudTreeView" renamed to "MudTreeView1". The compile error was one of the "MudTreeView1" left over.

tungi52 commented:
The new `BodyContent` would be nice, if someone need a fully customizable content for a `MudTreeItem`. Also I understand why the `Reload` method is necessary, at the moment there is no way to update children from server. I'm only not sure about the naming, on the other hand it is ok. However ReloadParent seems a bit meaningless to me, can you explain, why it is necessary? Also some tests would be great.

tungi52 commented on src/MudBlazor/Components/TreeView/MudTreeViewItem.razor.cs:
```diff
@@ -133,6 +133,13 @@ public bool Disabled
+        /// </summary>
+        [Parameter]
+        [Category(CategoryTypes.TreeView.Behavior)]
+        public RenderFragment<MudTreeViewItem<T>> BodyContent { get; set; }
```
Can you show/add an example why this approach is better  or provides more functionality than the existing `Content` `RenderFragment`? [https://mudblazor.com/components/treeview#custom-tree](https://mudblazor.com/components/treeview#custom-tree)

rpc-scandinavia replied:
At the end is my code, I create a tree from DataRecord objects.
I want to have edit buttons to the right of the text in each `TreeViewItem`.

### Why BodyContent
If I use `Content`, I have to handle the expansion/collapse and the icon my self;

From "[https://mudblazor.com/components/treeview#custom-tree](https://mudblazor.com/components/treeview#custom-tree)":

```
<Content>
	<MudTreeViewItemToggleButton @bind-Expanded="@context.IsExpanded" Visible="@context.HasChild" />
	<MudIcon Icon="@context.Icon" Class="ml-0 mr-2" Color="@Color.Default" />
	<MudText>@context.Text</MudText>
</Content>
```
In my code, `@context` is my DataRecord object, and I would really like to avoid the need to write my own tree logic, add properties to my DataRecord or wrap it in another class - but instead just use the logic already in the `TreeViewItem`.

### Why Reload and ReloadParent
Yes, the names can be something else.
When my TreeViewItem's DataRecord is modified, deleted or another is added, I want to update the tree. I added the `ReloadParent` to have a way of reloading/updating the child nodes in a `TreeViewItem`.

My "ItemEditButtons" have two event handlers, which gets the `TreeViewItem` from `@Content`, and calls the `ReloadParent` method.

Notice that the `BodyContent` `RenderFragment` gets the `TreeViewItem` as its content:

`RenderFragment<MudTreeView1Item<T>> BodyContent`


### My code
Notice that I added a "1" after "MudTreeView" in the class names in my project.

```
<MudPaper Elevation="0" Style="@this.elementStyle.ToString()">
	<MudTreeView1 ServerData="@this.LoadChildRecordsAsync" Items="@this.treeRecords" Hover="true">
		<ItemTemplate
			Context="dataRecord"
		>
			<MudTreeView1Item
				Value="@dataRecord"
				Icon="@this.GetRecordIcon(dataRecord)"
				IconColor="@this.GetRecordIconColor(dataRecord)"
				LoadingIconColor="Color.Info"
				Text="@dataRecord.Name"
			>
				<BodyContent>
					<div Style="@this.bodyContentStyle.ToString()">
						<MudText Style="@this.bodyTextStyle.ToString()">@dataRecord.Name</MudText>
						<div Style="@this.bodyButtonsStyle.ToString()">
							<MudIconButton Icon="@Icons.Filled.Autorenew" Size="Size.Medium" Color="Color.Inherit" OnClick="@(() => this.HandleRecordUpdated(context, dataRecord, dataRecord))" />

							<ItemEditButtons
								DataRecord="@dataRecord"
								DataRecordSaved="@((savedRecord) => this.HandleRecordUpdated(context, dataRecord, savedRecord))"
								DataRecordDeleted="@((deletedRecord) => this.HandleRecordUpdated(context, dataRecord, deletedRecord))"
							/>
						</div>
					</div>
				</BodyContent>
			</MudTreeView1Item>
		</ItemTemplate>
	</MudTreeView1>
</MudPaper>

```

rpc-scandinavia replied:
@tungi52: However ReloadParent seems a bit meaningless to me, can you explain, why it is necessary?

When I add sibling nodes, edit or remove nodes in my tree and database, I do not have a reference to the parent `MudTreeViewItem`, however the `MudTreeViewItem` does.

The `Parent` property in the `MudTreeViewItem` is private, so I can not traverse up the tree my self.

By the way, another problem I have, is to auto-expand the tree, so the selected node is visible (perhaps not scrolled into view, but the path to it is expanded) - but I have not been looking at that problem in detail.

tungi52 replied:
@rpc-scandinavia `BodyContent `and `Reload `is clear, but still have problem with the `ReloadParent`. This doesn't refresh the tree recursively, only has effect on the parent item. This is too specific to me.

> When I add sibling nodes, edit or remove nodes in my tree and database

But this could happen anywhere in the tree isn't it? Wouldn't be better if MudTree has a reload functionality, and recursively refresh the tree (maybe only the expanded nodes)?

rpc-scandinavia replied:
`ReloadParent` is necessary because it is the parent that loads and creates/shows the `MudTreeViewItem`'s

If i call `Reload` on the `MudTreeViewItem` that should be updated, it does not reload the specific item, IT reloads ITS children.

So to update the item, I use ReloadParent which reloads ITS children, including the item I want to reload.

```
a1
   b1
   b2
      c1
      c2
a2
```
If `b2` is updated and needs a reload, I call reload on `a1` because it clears its children (`b1` and `b2`) and reloads them, including `b2`.
If I call reload on `b2`, it clears its children (`c1` and `c2`) and reloads them.

As you can see, the `Reload` method clear its items, and loads them from the server again:

```
        public void Reload()
        {
			if (Items != null)
			{
				Items.Clear();
			}
			TryInvokeServerLoadFunc();
        }

```

rpc-scandinavia replied:
_Wouldn't be better if MudTree has a reload functionality, and recursively refresh the tree (maybe only the expanded nodes)?_

Perhaps that could be another feature. But remember that the clearing of the child items and reload, momentarily removes the child items on the screen - so I would create some "flashing" if all expanded items reloads. In this way, only the affected note "flashes" as it reloads.

However, if it was possible to just have a parent reload a specific item from the server, and update that item, there would be no flashing.

tungi52 replied:
Ok. So basically when you call `Reload`, you also want to update not only the children, but the treeitem itself. Sorry, but personally I still don't think, that this `ReloadParent` approach is a good solution. For me it seems too specific. What about these two other options?
- In the reload, after calling `ServerLoadFunc`, you can add basically what you have in the `ReloadParent`, like this:

```
if(Parent != null) 
{
  Parent.StateHasChanged();
}
else if(MudTreeRoot != null)
{
  MudTreeRoot.CallStateHasChanged();
}
```

   In this case you should update the treeitem's context in the ServerLoadFunc.
- Make `Parent` readable outside the component.

rpc-scandinavia replied:
I have made the changes you suggested.

In my project, I made another way of updating the "parent" in  my event handling - so I did not make the `Parent` public in my recent modifications.

mikes-gh commented:
I know this is still in discussion but `BodyContent` should be `ChildContent` in order to be consistent with the rest of the library

rpc-scandinavia commented:
> I know this is still in discussion but `BodyContent` should be `ChildContent` in order to be consistent with the rest of the library

I just wanted to make that change, but realised that `ChildContent` is already used.

tungi52 commented:
Thanks! For me this is ok. Can you add test(s) and an example to demonstrate the new features?

rpc-scandinavia commented:
> Thanks! For me this is ok. Can you add test(s) and an example to demonstrate the new features?

I will look at that.

tungi52 commented:
@rpc-scandinavia This doesn't build for me.

tungi52 commented on src/MudBlazor.UnitTests.Viewer/TestComponents/TreeView/TreeViewTest3.razor:
```diff
@@ -0,0 +1,7 @@
+<MudTreeView T="string">
```
MIssing namespace: `@namespace MudBlazor.UnitTests.TestComponents`

tungi52 commented on src/MudBlazor.UnitTests/Components/TreeViewTest.cs:
```diff
@@ -108,5 +108,13 @@ public void TreeViewServerTest()
+        public void TreeViewBodyContentTest()
+        {
+            var comp = Context.RenderComponent<TreeViewTest3>();
+            //Console.WriteLine(comp.Markup);
```
What is the purpose of this test? I think you could test the reload functionality, that the server changes will appear on the client side.

tungi52 commented on src/MudBlazor/Components/TreeView/MudTreeViewItem.razor.cs:
```diff
@@ -131,7 +131,14 @@ public bool Disabled
         [Parameter]
         [Category(CategoryTypes.TreeView.Behavior)]
-        public RenderFragment Content { get; set; }
+        public RenderFragment<MudTreeViewItem<T>> Content { get; set; }
```
You changed `RenderFragment` to `RenderFragment<MudTreeViewItem<T>>`. This is a breaking change, also affects `TreeViewCustomCheckboxExample`. Is this necessary? (not misunderstand me, I'm not against this, just curious and like to understand)

rpc-scandinavia replied:
You are right, I was not thinking.

rpc-scandinavia replied:
I have changed the code.
Options:

1) Do nothing to `Content`.
2) Make the breaking change to `Content` as I originally did.
3) Add the `ContentWithContext` as I have done in the code.

With 3) you can consider if `BodyContent` should be named `BodyContentWithContext` for consistency.

Off course 3) is a little awkward, with two Content's, but no breaking change :-)

ScarletKuro commented:
I didn't see this PR when I was creating this isssue https://github.com/MudBlazor/MudBlazor/issues/4503 but it looks like this changes should also solve the problems described in the ticket. I will leave the ticket open tho, until this is merged.

soend commented:
Anything else stopping this PR to get accepted and merged? I also need a way to use ServerData with custom MudTreeViewItem content. This PR would solve that issue.

just-the-benno commented:
I'm sorry that it took as so long to get back. 

@rpc-scandinavia, is there still life in this PR? Can we resurrect it?

rpc-scandinavia commented:
> is there still life in this PR? Can we resurrect it?

Well, everything worked, the only thing missing was the test - and I don't know how to make the test I want. Anyway, there has been so many changes at your end, that I can't get my GitHub copy to work - which probably mean that I have to delete my copy, fork it again, make all the changes again, and create a new pull request. So much hazzle for around 10 lines of code.

I have been using my solution since the first post, and I think we ended up with something good. There was however this question:

> With 3) you can consider if BodyContent should be named BodyContentWithContext for consistency.

I only want to spend more time on this GitHub pull, if you put it into the project.

rpc-scandinavia committed:
Update MudTreeView.razor.cs

rpc-scandinavia committed:
Update MudTreeViewItem.razor

rpc-scandinavia committed:
Update MudTreeViewItem.razor.cs

rpc-scandinavia commented:
Deleted my fork, and forked the project again.

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

ScarletKuro commented:
@henon what should be done for this to be merged? I think this is important PR to be able customize MudTreeViewItem.
Do we need a test or expand the docs?

henon commented:
Yes, we need test coverage of the new feature. A simple doc example would be nice to have also.

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

rpc-scandinavia committed:
MudTreeViewItem: Added 'BodyContent' which renders insted of text, end text and end icon.

Simple example.

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

rpc-scandinavia committed:
Update TreeViewBodyContentExample.razor

henon commented on src/MudBlazor/Components/TreeView/MudTreeViewItem.razor:
```diff
@@ -14,33 +14,40 @@
+			}
+			else
+			{
+				<MudText Typo="@TextTypo" Class="@TextClassname">
+					@Text
+				</MudText>
 
-            @if (!string.IsNullOrEmpty(EndIcon))
-            {
-                <div class="mud-treeview-item-icon">
-                    <MudIcon Icon="@EndIcon" Color="@EndIconColor" />
-                </div>
-            }
+				@if (!string.IsNullOrEmpty(EndText))
+				{
+					<MudText Typo="@EndTextTypo" Class="@EndTextClass">
+						@EndText
+					</MudText>
+				}
+
+				@if (!string.IsNullOrEmpty(EndIcon))
+				{
+					<div class="mud-treeview-item-icon">
+						<MudIcon Icon="@EndIcon" Color="@EndIconColor" />
+					</div>
+				}
+			}
```
There is some weird reformatting going on here and in other parts of the file. Can you please change it back to four spaces of indentation instead of tabs?

rpc-scandinavia replied:
Yes, sorry I have set my Codium to correct TABS.

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

rpc-scandinavia committed:
Changed whitespaces

rpc-scandinavia committed:
Changed whitespaces

rpc-scandinavia committed:
Changed whitespaces

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

henon commented:
@rpc-scandinavia Are you still adding tests? Because that is the only thing keeping this PR from being merged.

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

rpc-scandinavia committed:
Example with child items.

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
BodyContent

rpc-scandinavia commented:
@henon, I don't know how to make the correct tests. I have updated the example and added the place for the test, but I don't know how to test that the rendered HTML is as expected.

I can see in some of the other that they search and count, but the rendered tree must be correctly tested, surely it isn't enough just to count HTML elements of each type.

When I do "dotnet test" on my machine, if fails the fourth "Failed TimePicker_Should_Clear [50 ms]" test.

rpc-scandinavia committed:
Custom body content

henon commented on src/MudBlazor.Docs/Pages/Components/TreeView/TreeViewPage.razor:
```diff
@@ -129,6 +129,18 @@
+        <DocsPageSection>
+            <SectionHeader Title="Custom body content">
+                <Description>
+                    Use the <CodeInline>Body Content</CodeInline> insted of the <CodeInline>Content</CodeInline> render fragment,
```
`instead` instead of `insted`

henon commented on src/MudBlazor.Docs/Pages/Components/TreeView/TreeViewPage.razor:
```diff
@@ -129,6 +129,18 @@
+            <SectionHeader Title="Custom body content">
+                <Description>
+                    Use the <CodeInline>Body Content</CodeInline> insted of the <CodeInline>Content</CodeInline> render fragment,
+                    when you want your custom tree items, but still use the buildin icons and expansion buttons.
```
`built-in`

henon commented:
I think you are overthinking this. We can assume that Blazor renders RenderFragments correctly if it renders them at all.

The purpose of the test is to make sure the feature you added works. If someone would later change the razor file in a way that the custom body content is not rendered any more the test should fail.  So my suggestion is that you do a check for some unique html tag that can only be added via the new custom BodyContent render fragment.

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

rpc-scandinavia commented:
Ok @henon, I corrected the spelling errors and added some simple tests.

rpc-scandinavia commented:
@henon, another thing in `MudTreeViewItem.razor.cs`, I had to replace all:

`StateHasChanged();`

With:

`InvokeAsync(() => StateHasChanged());`

Because of:

`Unhandled exception. System.InvalidOperationException: The current thread is not associated with the Dispatcher. Use InvokeAsync() to switch execution to the Dispatcher when triggering rendering or component state.`

Do you think that should be done in this PR ?

ScarletKuro commented:
I wonder if it's possible to test the new Reload method that was introduced.

> another thing in MudTreeViewItem.razor.cs, I had to replace all:

Can you be more specific where, there are many of them
But if the method is Task, i would propose `await InvokeAsync(StateHasChanged);`

FYI: I think github it dead and you can't push rn.

ScarletKuro commented:
Not related to this PR, but why is `TryInvokeServerLoadFunc` an async **void** instead of async **Task**. I see no reason for it and it's always called from a Task and you would be able to await it.
Why `_childItems.ForEach(async c => await c.SelectItem(value, source));` instead of 
```Csharp
foeach (var items in _childItems){
    await items .SelectItem(value, source)
}
```
I recommend to avoid async void unless totally necessary i.e. in event handler, for example.

I kinda want to propose to make Reload->ReloadAsync with an Task since MudTree is has async semantics. the `MudTreeRoot.ServerData` is async and it would allow me to make a PR with some inner change to improve the component without breaking the public API.

henon commented:
Yes and of course the ReloadAsync function can also be tested easily. I.e. by changin the server data before calling it and checking that the changed data is reflected in the tree.

henon commented:
> Do you think that should be done in this PR ?

Only if you are sure it is necessary. The first question to ask is why were they called from the wrong thread. What caused this?

ScarletKuro commented:
> Only if you are sure it is necessary. The first question to ask is why were they called from the wrong thread. What caused this?

It's only my guess, but maybe he modified some state in a test, like changing value of [parameter] property.
Those should be called like this in a test
`await comp.InvokeAsync(() => item.Instance.Disabled = true);`
`await comp.InvokeAsync(() => comp.Instance.SelectFirst());`

ScarletKuro commented:
> I saw that as well

Ok, but lets keep it as is in this PR.

Please, change `void Reload` to `void ReloadAsync` and add test for this method like henon suggested:

>Yes and of course the ReloadAsync function can also be tested easily. I.e. by changin the server data before calling it and checking that the changed data is reflected in the tree.

And then I guess we are good to go and merge this PR.

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

ScarletKuro commented on src/MudBlazor/Components/TreeView/MudTreeViewItem.razor.cs:
```diff
@@ -351,21 +358,40 @@ protected async Task OnItemDoubleClicked(MouseEventArgs ev)
+            }
+            else if (MudTreeRoot != null)
+            {
+                MudTreeRoot.CallStateHasChanged();
```
If you rebase to latest changes there is new interface now and you can do something like this
`((IMudStateHasChanged)MudTreeRoot).StateHasChanged()` and remove the `CallStateHasChanged`

rpc-scandinavia committed:
Merge branch 'MudBlazor:dev' into dev

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

rpc-scandinavia committed:
Custom body content

rpc-scandinavia commented:
@ScarletKuro, I have made the `async` and `IMudStateHasChanged` changes.

ScarletKuro commented:
lgtm

henon commented:
Thanks for your contribution @rpc-scandinavia !
